File: solutions/2026-05-09/cpp/2026-05-09-cyclically-rotating-a-grid.cpp
```cpp
// Problem: Cyclically Rotating a Grid
// Problem Link: https://leetcode.com/problems/cyclically-rotating-a-grid/
// Approach:
// The core idea is to treat each layer of the grid as a separate 1D list.
// We can extract the elements of a layer in counter-clockwise order,
// perform the cyclic rotation on this 1D list, and then place the rotated
// elements back into the grid.
//
// To extract a layer:
// Iterate from the top-left corner (r1, c1) to the top-right (r1, c2),
// then to the bottom-right (r2, c2), then to the bottom-left (r2, c1),
// and finally back to the start (r1, c1) excluding the first element to avoid duplicates.
//
// To rotate the 1D list:
// A cyclic rotation of k elements to the left on a list of size L can be achieved
// by taking the last L - (k % L) elements and placing them at the beginning,
// followed by the first (k % L) elements. This is equivalent to a k % L left shift.
//
// To place the rotated elements back:
// Iterate through the layer's perimeter again in the same counter-clockwise
// order and populate the grid with elements from the rotated 1D list.
//
// The process is repeated for all layers, from the outermost to the innermost.
//
// Time Complexity:
// Let m be the number of rows and n be the number of columns.
// The number of layers is min(m, n) / 2.
// For each layer, we traverse its perimeter. The perimeter of a layer at depth 'd'
// (0-indexed from the outside) has a length of 2 * (m - 2d - 1) + 2 * (n - 2d - 1).
// The total number of elements in the grid is m * n.
// Extracting elements from a layer takes O(perimeter_length).
// Rotating the 1D list takes O(perimeter_length).
// Placing elements back takes O(perimeter_length).
// The total time for one rotation is O(m*n) as we visit each element a constant number of times.
// Since we need to perform k rotations, and k can be very large, we need to optimize.
// The number of distinct states for a layer is equal to its perimeter length.
// So, we only need to perform k % perimeter_length rotations for each layer.
// Thus, the effective rotation for a layer is k_eff = k % perimeter_length.
// The total time complexity is O(m * n), as each element is processed a constant number of times for all layers and rotations.
//
// Space Complexity:
// We use a temporary vector to store the elements of each layer. The maximum size of this vector
// is the perimeter of the outermost layer, which is 2*(m-1) + 2*(n-1).
// Therefore, the space complexity is O(m + n).
//
// Example 1:
// grid = [[40,10],[30,20]], k = 1
// Layer 0: [40, 10, 20, 30] (counter-clockwise)
// Perimeter length = 4. k_eff = 1 % 4 = 1.
// Rotate [40, 10, 20, 30] by 1 to the left: [10, 20, 30, 40]
// Place back:
// grid[0][0] = 10, grid[0][1] = 20
// grid[1][1] = 30, grid[1][0] = 40
// Output: [[10,20],[40,30]]
//
// Example 2:
// grid = [[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]], k = 2
// Layer 0: [1, 2, 3, 4, 8, 12, 16, 15, 14, 13, 9, 5] (counter-clockwise)
// Perimeter length = 12. k_eff = 2 % 12 = 2.
// Rotate [1, 2, 3, 4, 8, 12, 16, 15, 14, 13, 9, 5] by 2 to the left:
// [3, 4, 8, 12, 16, 15, 14, 13, 9, 5, 1, 2]
// Place back:
// grid[0][0]=3, grid[0][1]=4, grid[0][2]=8, grid[0][3]=12
// grid[1][3]=16, grid[2][3]=15, grid[3][3]=14
// grid[3][2]=13, grid[3][1]=9, grid[3][0]=5
// grid[2][0]=1, grid[1][0]=7 (Mistake in manual trace, should be grid[1][0] = 2)
// Corrected:
// grid[0][0]=3, grid[0][1]=4, grid[0][2]=8, grid[0][3]=12
// grid[1][3]=16, grid[2][3]=15, grid[3][3]=14
// grid[3][2]=13, grid[3][1]=9, grid[3][0]=5
// grid[2][0]=1, grid[1][0]=2
// For inner layer:
// Layer 1: [6, 7, 10, 11]
// Perimeter length = 4. k_eff = 2 % 4 = 2.
// Rotate [6, 7, 10, 11] by 2 to the left: [10, 11, 6, 7]
// Place back:
// grid[1][1]=10, grid[1][2]=11
// grid[2][2]=6, grid[2][1]=7
// Final: [[3,4,8,12],[2,11,10,16],[1,7,6,15],[5,9,13,14]]
// The explanation has a small mistake in manual trace.
// The provided logic will work correctly.
//
// Important Note: m and n are guaranteed to be even.
//
#include <vector>
#include <algorithm>

class Solution {
public:
    // Function to cyclically rotate the grid k times.
    std::vector<std::vector<int>> rotateGrid(std::vector<std::vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();

        // Iterate through each layer of the grid.
        // Layers are defined by their distance from the outer boundary.
        // The number of layers is min(m, n) / 2.
        // r1, c1 are the top-left coordinates of the current layer.
        // r2, c2 are the bottom-right coordinates of the current layer.
        for (int r1 = 0, c1 = 0, r2 = m - 1, c2 = n - 1; r1 < r2 && c1 < c2; ++r1, ++c1, --r2, --c2) {
            // Extract the elements of the current layer into a 1D vector.
            std::vector<int> layer_elements;

            // Traverse top row from left to right
            for (int j = c1; j <= c2; ++j) {
                layer_elements.push_back(grid[r1][j]);
            }
            // Traverse right column from top to bottom (excluding the corner already added)
            for (int i = r1 + 1; i <= r2; ++i) {
                layer_elements.push_back(grid[i][c2]);
            }
            // Traverse bottom row from right to left (excluding the corner already added)
            for (int j = c2 - 1; j >= c1; --j) {
                layer_elements.push_back(grid[r2][j]);
            }
            // Traverse left column from bottom to top (excluding the corner already added)
            for (int i = r2 - 1; i > r1; --i) {
                layer_elements.push_back(grid[i][c1]);
            }

            // Calculate the effective number of rotations needed.
            // The cycle length of rotations for a layer is equal to its perimeter length.
            int layer_perimeter = layer_elements.size();
            if (layer_perimeter == 0) continue; // Should not happen for valid grids and layers
            int effective_k = k % layer_perimeter;

            // If effective_k is 0, no rotation is needed for this layer.
            if (effective_k == 0) continue;

            // Perform the cyclic rotation on the 1D vector.
            // This is a left rotation by effective_k positions.
            // std::rotate shifts the element at `first + n` to the beginning.
            // So, to rotate left by `effective_k`, we specify the element at `layer_elements.begin() + effective_k`.
            std::rotate(layer_elements.begin(), layer_elements.begin() + effective_k, layer_elements.end());

            // Place the rotated elements back into the grid.
            int element_index = 0;
            // Traverse top row from left to right
            for (int j = c1; j <= c2; ++j) {
                grid[r1][j] = layer_elements[element_index++];
            }
            // Traverse right column from top to bottom
            for (int i = r1 + 1; i <= r2; ++i) {
                grid[i][c2] = layer_elements[element_index++];
            }
            // Traverse bottom row from right to left
            for (int j = c2 - 1; j >= c1; --j) {
                grid[r2][j] = layer_elements[element_index++];
            }
            // Traverse left column from bottom to top
            for (int i = r2 - 1; i > r1; --i) {
                grid[i][c1] = layer_elements[element_index++];
            }
        }

        return grid;
    }
};
```

File: solutions/2026-05-09/python/2026-05-09-cyclically-rotating-a-grid.py
```python
# Problem: Cyclically Rotating a Grid
# Link: https://leetcode.com/problems/cyclically-rotating-a-grid/
#
# Approach:
# The problem requires cyclically rotating each layer of the grid counter-clockwise.
# We can extract each layer as a 1D list, perform the cyclic rotation on this list,
# and then place the rotated elements back into the grid.
#
# To extract layers:
# A layer is defined by its top-left corner (row1, col1) and bottom-right corner (row2, col2).
# For a grid of size m x n, the outermost layer has corners (0, 0) and (m-1, n-1).
# The next layer has corners (1, 1) and (m-2, n-2), and so on.
# We can iterate from the outermost layer inwards. For each layer, we collect its elements
# by traversing its perimeter: top row (left to right), right column (top to bottom),
# bottom row (right to left), and left column (bottom to top).
#
# Cyclic Rotation of a 1D list:
# To cyclically rotate a list `arr` by `k` positions counter-clockwise, we can use the
# modulo operator. If the length of the list is `L`, then `k` rotations is equivalent to
# `k % L` rotations. The new position of an element at index `i` will be `(i - k % L + L) % L`.
# A simpler way is to slice the list: `rotated_arr = arr[k % L:] + arr[:k % L]`.
#
# Placing elements back:
# After rotating the 1D list for a layer, we place these elements back into the grid
# in the same order they were extracted, effectively performing the counter-clockwise
# rotation.
#
# The number of layers is `min(m, n) // 2`.
# We iterate through these layers. For each layer, we determine its boundary coordinates.
# We then extract the elements of that layer into a temporary list.
# We perform the cyclic rotation on this temporary list.
# Finally, we populate the grid with the rotated elements.
#
# Time Complexity:
# Let m be the number of rows and n be the number of columns.
# Extracting elements for a layer takes O(perimeter of layer) time. The perimeter of layer `l` is 2*(m - 2*l) + 2*(n - 2*l) - 4.
# Summing the perimeters for all layers:
# For the outermost layer (l=0), perimeter is 2*m + 2*n - 4.
# For the next layer (l=1), perimeter is 2*(m-2) + 2*(n-2) - 4.
# ...
# The total number of elements in the grid is m*n.
# Extracting all elements and placing them back takes O(m*n) time.
# Performing cyclic rotation on a list of size `P` (perimeter) takes O(P) time using slicing.
# The sum of perimeters is proportional to m*n.
# Thus, the total time complexity is O(m*n).
#
# Space Complexity:
# We use a temporary list to store the elements of a single layer. The maximum size of this
# list is the perimeter of the largest layer, which is O(m+n).
# Therefore, the space complexity is O(m+n).
#
class Solution:
    def cyclicallyRotatingGrid(self, grid: list[list[int]], k: int) -> list[list[int]]:
        m = len(grid)
        n = len(grid[0])
        
        # Number of layers in the grid
        num_layers = min(m, n) // 2

        # Iterate through each layer from outermost to innermost
        for layer_idx in range(num_layers):
            # Define the boundaries of the current layer
            row1, col1 = layer_idx, layer_idx
            row2, col2 = m - 1 - layer_idx, n - 1 - layer_idx

            # If the layer is invalid (e.g., a single row/column which shouldn't happen with even m, n)
            if row1 >= row2 or col1 >= col2:
                break
            
            # Extract elements of the current layer into a 1D list
            layer_elements = []
            
            # Traverse top row (left to right)
            for c in range(col1, col2):
                layer_elements.append(grid[row1][c])
            
            # Traverse right column (top to bottom)
            for r in range(row1, row2):
                layer_elements.append(grid[r][col2])
            
            # Traverse bottom row (right to left)
            for c in range(col2, col1, -1):
                layer_elements.append(grid[row2][c])
            
            # Traverse left column (bottom to top)
            for r in range(row2, row1, -1):
                layer_elements.append(grid[r][col1])

            # Calculate the effective number of rotations for this layer
            # The length of the layer elements list is its perimeter
            layer_len = len(layer_elements)
            effective_k = k % layer_len
            
            # Perform the cyclic rotation on the extracted layer elements
            # Counter-clockwise rotation: elements from index `effective_k` to the end
            # followed by elements from the beginning up to `effective_k`.
            rotated_layer_elements = layer_elements[effective_k:] + layer_elements[:effective_k]
            
            # Place the rotated elements back into the grid for the current layer
            element_idx = 0
            
            # Traverse top row (left to right)
            for c in range(col1, col2):
                grid[row1][c] = rotated_layer_elements[element_idx]
                element_idx += 1
            
            # Traverse right column (top to bottom)
            for r in range(row1, row2):
                grid[r][col2] = rotated_layer_elements[element_idx]
                element_idx += 1
            
            # Traverse bottom row (right to left)
            for c in range(col2, col1, -1):
                grid[row2][c] = rotated_layer_elements[element_idx]
                element_idx += 1
            
            # Traverse left column (bottom to top)
            for r in range(row2, row1, -1):
                grid[r][col1] = rotated_layer_elements[element_idx]
                element_idx += 1
        
        return grid

```